library/psa_crypto.c

```diff
@@ -3425,25 +3335,38 @@ psa_status_t psa_cipher_generate_iv( psa_cipher_operation_t *operation,
                                      size_t iv_size,
                                      size_t *iv_length )
 {
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
 
+    *iv_length = 0;
+
     if( operation->id == 0 )
     {
         return( PSA_ERROR_BAD_STATE );
     }
 
+    if( iv_size < operation->iv_size )
+        return( PSA_ERROR_BUFFER_TOO_SMALL );
+
     if( operation->iv_set || ! operation->iv_required )
     {
         return( PSA_ERROR_BAD_STATE );
     }
 
-    status = psa_driver_wrapper_cipher_generate_iv( operation,
-                                                    iv,
-                                                    iv_size,
-                                                    iv_length );
+    ret = mbedtls_psa_get_random( MBEDTLS_PSA_RANDOM_STATE, iv,
```

Comment:
Please leverage the PSA primitive `psa_generate_random` instead.

Reply:
First, this request seems a bit off of topic to me. Second I am not sure what we would gain anyway. What would be the benefit to use here psa_generate_random() instead of mbedtls_psa_get_random()? I can see that if  MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG is defined mbedtls_psa_get_random eventually calls  mbedtls_psa_external_get_random() thus the case of the external random generator is taken into account.

Reply:
The benefit is actually distinguishing between the different API usages.

`mbedtls_psa_get_random()` is meant to be passed as the `f_rng` argument to Mbed TLS library functions. In the PSA Core, `psa_get_random()` is the PSA-way of getting random bytes. The cipher_generate_iv function is currently the only place where `mbedtls_psa_get_random` is used standalone out of its Mbed TLS library function context.

`psa_generate_random` is already called inside the core for raw key generation, when there's no Mbed TLS library function involved. I suggest sticking to that distinction in order to not confuse people who are new to the codebase.

It'll make the implementation of cipher_generate_iv cleaner as well, since you don't need to deal with different status return types anymore.

Reply:
I agree it's only tangentially related to the changes in this PR, but if the function body is being updated already, I'm a firm believer in 'leave things in a better shape than you found them'.

Reply:
ok, thanks for the clarification. I didn't realize before how peculiar this call was. As it seems to be the only one in PSA crypto code and as I am moving it, let's improve it as well.